src/psyclone/domain/nemo/transformations/nemo_arrayrange2loop_trans.py

```diff
@@ -308,16 +302,71 @@ def validate(self, node, options=None):
+            # Allow non elemental UBOUND and LBOUND
+            if operation.operator is BinaryOperation.Operator.LBOUND:
+                continue
+            if operation.operator is BinaryOperation.Operator.UBOUND:
+                continue
+            if not operation.is_elemental():
                 raise TransformationError(
                     "Error in NemoArrayRange2LoopTrans transformation. This "
                     "transformation does not support array valued operations "
                     "on the rhs of the associated Assignment node, but found "
                     "'{0}'.".format(operation.operator.name))
+        # Do not allow to transform expressions with CodeBlocks
+        if assignment.walk(CodeBlock):
+            raise TransformationError(
+                "Error in NemoArrayRange2LoopTrans transformation. This "
+                "transformation does not support array assignments that "
+                "contain a CodeBlock anywhere in the expression.")
+        # Do not allow to transform expressions with function calls (to allow
+        # this we need to differentiate between elemental and not elemental
+        # functions as they have different semantics in array notation)
+        if assignment.walk(Call):
+            raise TransformationError(
+                "Error in NemoArrayRange2LoopTrans transformation. This "
+                "transformation does not support array assignments that "
+                "contain a Call anywhere in the expression.")
+
+        for reference in assignment.walk(Reference):
+            # As special case we always allow references to whole arrays as
+            # part of the LBOUND and UBOUND operations, regardless of the
+            # restrictions below (e.g. is a DeferredType reference).
+            if isinstance(reference.parent, Operation):
+                operator = reference.parent.operator
+                if operator is BinaryOperation.Operator.LBOUND:
+                    continue
+                if operator is BinaryOperation.Operator.UBOUND:
+                    continue
+
+            # We allow any references that are part of a structure syntax
+            if isinstance(reference, (StructureReference, StructureMember)):
+                continue
+
+            # We allow any references that have explicit array dimensions
+            if isinstance(reference, (ArrayReference, ArrayMember)):
+                # Arrays must have at least 1 range child in order to be
+                # converted to an explicit expression
+                if not any(child for child in reference.children if
+                           isinstance(child, Range)):
+                    raise TransformationError(
+                        "Error in NemoArrayRange2LoopTrans transformation. "
+                        "This transformation does not support assignments "
+                        "with rhs arrays that don't have a range.")
```

Comment:
Please give more details, e.g. the name of the array and the Fortran for the assignment.

Reply:
Done